File: src/shared/noteTags.ts
/** 单个标签最长字符数，避免把整段误粘贴进标签。 */
export const MAX_NOTE_TAG_CHARS = 40;

/** 可视化编辑时的标签数量上限，防止文末标签行过长。 */
export const MAX_NOTE_TAGS = 20;

/** 从 Markdown 正文提取标签，规则与后端 `storage::extract_tags` 对齐。 */
export function extractNoteTags(content: string): string[] {
  const tags = [...parseFrontmatterTags(content), ...parseHashtagLineTags(content)];

  return uniqueSortedTags(tags);
}

/** 把当前标签写回正文：已有 frontmatter `tags` 时改 YAML，否则维护独立 `#标签` 行。 */
export function applyNoteTags(content: string, nextTags: string[]): string {
  const normalized = uniqueSortedTags(nextTags);
  const frontmatter = splitFrontmatter(content);

  if (frontmatter && hasFrontmatterTagsKey(frontmatter.yaml)) {
    const yaml = upsertFrontmatterTags(frontmatter.yaml, normalized);
    const body = stripHashtagOnlyLines(frontmatter.body, 0);

    return formatFrontmatterDocument(yaml, body);
  }

  const stripped = stripHashtagOnlyLines(content, frontmatterBodyStartLine(content.split("\n")));

  return appendTrailingHashtags(stripped, normalized);
}

/** 去掉空值、非法字符和重复项，供输入框和正文解析共用。 */
export function normalizeTagName(raw: string): string | null {
  let name = raw.trim();

  if (name.startsWith("#") && !name.startsWith("##")) {
    name = name.slice(1).trim();
  }

  name = name.replace(/^[,\.;:!?，。；：！？"'“”]+|[,\.;:!?，。；：！？"'“”]+$/g, "").trim();

  if (!name || [...name].length > MAX_NOTE_TAG_CHARS) {
    return null;
  }

  if (/[\s#[\]{},:]/.test(name)) {
    return null;
  }

  return name;
}

/** 去重并按中文习惯排序，保证信息栏和编辑器顺序稳定。 */
function uniqueSortedTags(tags: string[]): string[] {
  const unique = new Set<string>();

  for (const tag of tags) {
    const normalized = normalizeTagName(tag);

    if (normalized) {
      unique.add(normalized);
    }
  }

  return [...unique].sort((left, right) => left.localeCompare(right, "zh-CN"));
}

/** 拆出闭合 YAML frontmatter；未闭合则视为普通正文。 */
function splitFrontmatter(content: string): { yaml: string; body: string } | null {
  const normalized = content.replace(/^\uFEFF/, "");
  const lines = normalized.split("\n");

  if (trimLine(lines[0] ?? "") !== "---") {
    return null;
  }

  for (let index = 1; index < lines.length; index += 1) {
    if (trimLine(lines[index] ?? "") === "---") {
      return {
        yaml: lines.slice(1, index).join("\n"),
        body: lines.slice(index + 1).join("\n"),
      };
    }
  }

  return null;
}

/** 读取 frontmatter 中的 tags 字段。 */
function parseFrontmatterTags(content: string): string[] {
  const frontmatter = splitFrontmatter(content);

  return frontmatter ? parseFrontmatterTagsFromYaml(frontmatter.yaml) : [];
}

/** 解析 `tags: a, b`、`tags: [a, b]` 和多行 `- item`。 */
function parseFrontmatterTagsFromYaml(yaml: string): string[] {
  const lines = yaml.split("\n");
  const tags: string[] = [];

  for (let index = 0; index < lines.length; index += 1) {
    const line = stripCarriageReturn(lines[index] ?? "");

    if (!line.startsWith("tags:")) {
      continue;
    }

    const rest = line.slice("tags:".length).trim();

    if (rest) {
      tags.push(...parseInlineTagList(rest));
      continue;
    }

    index += 1;
    while (index < lines.length) {
      const itemLine = stripCarriageReturn(lines[index] ?? "");
      const item = parseYamlListItem(itemLine);

      if (item !== null) {
        tags.push(item);
        index += 1;
        continue;
      }

      if (!itemLine.trim()) {
        index += 1;
        continue;
      }

      break;
    }
  }

  return tags;
}

/** 顶层是否已经声明 `tags` 键，用来决定可视化编辑写 YAML 还是写 `#标签` 行。 */
function hasFrontmatterTagsKey(yaml: string): boolean {
  return yaml.split("\n").some((line) => stripCarriageReturn(line).startsWith("tags:"));
}

/** 更新或插入 frontmatter 的 tags 字段，其它键保持原样。 */
function upsertFrontmatterTags(yaml: string, tags: string[]): string {
  const lines = yaml.split("\n").map(stripCarriageReturn);
  const rendered = tags.length ? `tags: [${tags.join(", ")}]` : null;
  let start = -1;
  let end = -1;

  for (let index = 0; index < lines.length; index += 1) {
    if (!lines[index]?.startsWith("tags:")) {
      continue;
    }

    start = index;
    end = index + 1;
    if (!lines[index].slice("tags:".length).trim()) {
      while (end < lines.length && parseYamlListItem(lines[end] ?? "") !== null) {
        end += 1;
      }
    }
    break;
  }

  if (start >= 0) {
    const nextLines = [...lines];

    if (rendered) {
      nextLines.splice(start, end - start, rendered);
    } else {
      nextLines.splice(start, end - start);
    }

    return nextLines.join("\n");
  }

  if (!rendered) {
    return yaml;
  }

  const nextLines = [...lines];
  let insertAt = 0;

  while (insertAt < nextLines.length && !nextLines[insertAt]?.trim()) {
    insertAt += 1;
  }

  nextLines.splice(insertAt, 0, rendered);

  return nextLines.join("\n");
}

/** 读取缩进后的 YAML 列表项。 */
function parseYamlListItem(line: string): string | null {
  const trimmedEnd = line.trimEnd();

  if (!trimmedEnd.startsWith(" ") && !trimmedEnd.startsWith("\t")) {
    return null;
  }

  const rest = trimmedEnd.trim();
  if (!rest.startsWith("- ")) {
    return null;
  }

  return unquoteYamlScalar(rest.slice(2).trim());
}

/** 解析逗号分隔或 YAML 流式数组。 */
function parseInlineTagList(raw: string): string[] {
  const trimmed = raw.trim();
  const inner =
    trimmed.startsWith("[") && trimmed.endsWith("]") && trimmed.length >= 2
      ? trimmed.slice(1, -1)
      : trimmed;

  return inner
    .split(",")
    .map((part) => unquoteYamlScalar(part.trim()))
    .filter(Boolean);
}

/** 去掉 YAML 标量两端引号。 */
function unquoteYamlScalar(value: string): string {
  const trimmed = value.trim();

  if (
    trimmed.length >= 2 &&
    ((trimmed.startsWith('"') && trimmed.endsWith('"')) || (trimmed.startsWith("'") && trimmed.endsWith("'")))
  ) {
    return trimmed.slice(1, -1);
  }

  return trimmed;
}

/** 收集独立 `#标签` 行，跳过 frontmatter 和代码围栏。 */
function parseHashtagLineTags(content: string): string[] {
  const lines = content.split("\n");
  const inFence = codeFenceMask(lines);
  const start = frontmatterBodyStartLine(lines);
  const tags: string[] = [];

  for (let index = start; index < lines.length; index += 1) {
    if (inFence[index]) {
      continue;
    }

    const lineTags = parseHashtagOnlyLine(lines[index] ?? "");

    if (lineTags) {
      tags.push(...lineTags);
    }
  }

  return tags;
}

/** 正文起始行号；没有闭合 frontmatter 时从文件头开始。 */
function frontmatterBodyStartLine(lines: string[]): number {
  if (trimLine(lines[0] ?? "") !== "---") {
    return 0;
  }

  for (let index = 1; index < lines.length; index += 1) {
    if (trimLine(lines[index] ?? "") === "---") {
      return index + 1;
    }
  }

  return 0;
}

/** 标记每一行是否位于 fenced code 内。 */
function codeFenceMask(lines: string[]): boolean[] {
  let inFence = false;

  return lines.map((line) => {
    const trimmed = trimLine(line);

    if (trimmed.startsWith("```") || trimmed.startsWith("~~~")) {
      inFence = !inFence;
      return true;
    }

    return inFence;
  });
}

/** 整行都是 `#标签` 时返回标签，否则返回 null。 */
function parseHashtagOnlyLine(line: string): string[] | null {
  const trimmed = trimLine(line);

  if (!trimmed || isAtxHeading(trimmed)) {
    return null;
  }

  const tags: string[] = [];

  for (const token of trimmed.split(/\s+/)) {
    const name = hashtagTokenName(token);

    if (!name) {
      return null;
    }

    tags.push(name);
  }

  return tags.length ? tags : null;
}

/** CommonMark ATX 标题需要 `#` 后接空白。 */
function isAtxHeading(trimmed: string): boolean {
  const match = trimmed.match(/^(#{1,6})(.*)$/);

  if (!match) {
    return false;
  }

  const rest = match[2] ?? "";

  return rest.length === 0 || /^\s/.test(rest);
}

/** 解析单个 `#标签` token。 */
function hashtagTokenName(token: string): string | null {
  if (!token.startsWith("#") || token.startsWith("##")) {
    return null;
  }

  return normalizeTagName(token.slice(1));
}

/** 删除正文中的独立标签行，供可视化编辑重写前清理旧表示。 */
function stripHashtagOnlyLines(content: string, bodyStart: number): string {
  const lines = content.split("\n");
  const inFence = codeFenceMask(lines);
  const kept: string[] = [];

  for (let index = 0; index < lines.length; index += 1) {
    const line = lines[index] ?? "";
    const shouldStrip = index >= bodyStart && !inFence[index] && parseHashtagOnlyLine(line);

    if (shouldStrip) {
      continue;
    }

    kept.push(line);
  }

  return collapseExtraBlankLines(kept.join("\n")).replace(/\s+$/u, "");
}

/** 把标签写到正文末尾单独一行。 */
function appendTrailingHashtags(body: string, tags: string[]): string {
  const trimmed = body.replace(/\s+$/u, "");

  if (!tags.length) {
    return trimmed ? `${trimmed}\n` : "";
  }

  const tagLine = tags.map((tag) => `#${tag}`).join(" ");

  return trimmed ? `${trimmed}\n\n${tagLine}\n` : `${tagLine}\n`;
}

/** 用更新后的 YAML 重新拼出带 frontmatter 的文档；YAML 被删空时不再保留空壳。 */
function formatFrontmatterDocument(yaml: string, body: string): string {
  const yamlTrimmed = yaml.replace(/^\n+/u, "").replace(/\n+$/u, "");
  const bodyContent = body.replace(/^\n+/u, "").replace(/\s+$/u, "");

  if (!yamlTrimmed) {
    return bodyContent ? `${bodyContent}\n` : "";
  }

  return `---\n${yamlTrimmed}\n---\n${bodyContent ? `\n${bodyContent}\n` : ""}`;
}

/** 去掉标签行后把连续空行收成最多一行间隔。 */
function collapseExtraBlankLines(content: string): string {
  return content.replace(/\n{3,}/g, "\n\n");
}

/** 去掉行尾 CR，便于同时处理 Windows 换行。 */
function stripCarriageReturn(line: string): string {
  return line.endsWith("\r") ? line.slice(0, -1) : line;
}

/** 去掉 CR 后再 trim，用于识别 `---` 分隔符。 */
function trimLine(line: string): string {
  return stripCarriageReturn(line).trim();
}

File: src-tauri/src/storage/note_tags.rs
//! Markdown 笔记标签。
//!
//! 标签只来自文件正文，避免信息栏里的数量和正文脱节：
//! 1. 文首 YAML frontmatter 的 `tags` 字段；
//! 2. 不在代码围栏内、整行只由 `#标签` 组成的行（常见写法是文末一行）。
//!
//! 标题 `# 标题`、段落里的 `#词` 和代码注释都不会当成标签。

/// 单个标签最长字符数，避免把整段误粘贴进标签。
const MAX_TAG_CHARS: usize = 40;

/// 从 Markdown 正文提取去重后的标签，按 UTF-8 字节序排列。
pub fn extract_tags(content: &str) -> Vec<String> {
    let mut tags = parse_frontmatter_tags(content);
    tags.extend(parse_hashtag_line_tags(content));
    normalize_tags(tags)
}

/// 去掉空值、非法字符和重复项。
fn normalize_tags(tags: Vec<String>) -> Vec<String> {
    let mut normalized: Vec<String> = tags
        .into_iter()
        .filter_map(|tag| normalize_tag_name(&tag))
        .collect();
    normalized.sort();
    normalized.dedup();
    normalized
}

/// 把用户输入或 YAML 标量收成合法标签名；失败则丢弃。
fn normalize_tag_name(raw: &str) -> Option<String> {
    let mut name = raw.trim();
    if let Some(stripped) = name.strip_prefix('#') {
        if !stripped.starts_with('#') {
            name = stripped.trim();
        }
    }

    let name = name
        .trim_matches(|ch: char| {
            matches!(
                ch,
                ',' | '.'
                    | ';'
                    | ':'
                    | '!'
                    | '?'
                    | '，'
                    | '。'
                    | '；'
                    | '：'
                    | '！'
                    | '？'
                    | '"'
                    | '\''
                    | '“'
                    | '”'
            )
        })
        .trim();

    if name.is_empty() || name.chars().count() > MAX_TAG_CHARS {
        return None;
    }

    if name
        .chars()
        .any(|ch| ch.is_whitespace() || matches!(ch, '#' | '[' | ']' | '{' | '}' | ',' | ':'))
    {
        return None;
    }

    Some(name.to_owned())
}

/// 读取文首 YAML frontmatter 中的 tags 字段。
fn parse_frontmatter_tags(content: &str) -> Vec<String> {
    split_frontmatter(content)
        .map(|(yaml, _body)| parse_frontmatter_tags_from_yaml(&yaml))
        .unwrap_or_default()
}

/// 拆出闭合的 YAML frontmatter；未闭合或没有起始 `---` 时视为没有。
fn split_frontmatter(content: &str) -> Option<(String, String)> {
    let content = content.strip_prefix('\u{feff}').unwrap_or(content);
    let lines: Vec<&str> = content.split('\n').collect();
    if lines.first()?.trim_end_matches('\r').trim() != "---" {
        return None;
    }

    for (index, line) in lines.iter().enumerate().skip(1) {
        if line.trim_end_matches('\r').trim() == "---" {
            return Some((lines[1..index].join("\n"), lines[index + 1..].join("\n")));
        }
    }

    None
}

/// 解析 frontmatter 里的 `tags`，兼容数组、逗号列表和多行 `- item`。
fn parse_frontmatter_tags_from_yaml(yaml: &str) -> Vec<String> {
    let lines: Vec<&str> = yaml.split('\n').collect();
    let mut tags = Vec::new();
    let mut index = 0;

    while index < lines.len() {
        let line = lines[index].trim_end_matches('\r');
        if let Some(rest) = line.strip_prefix("tags:") {
            let rest = rest.trim();
            if rest.is_empty() {
                index += 1;
                while index < lines.len() {
                    let item_line = lines[index].trim_end_matches('\r');
                    if let Some(item) = parse_yaml_list_item(item_line) {
                        tags.push(item);
                        index += 1;
                    } else if item_line.trim().is_empty() {
                        index += 1;
                    } else {
                        break;
                    }
                }
            } else {
                tags.extend(parse_inline_tag_list(rest));
                index += 1;
            }
        } else {
            index += 1;
        }
    }

    tags
}

/// 读取缩进后的 YAML 列表项 `- value`。
fn parse_yaml_list_item(line: &str) -> Option<String> {
    let trimmed_end = line.trim_end();
    if !trimmed_end.starts_with(' ') && !trimmed_end.starts_with('\t') {
        return None;
    }

    let rest = trimmed_end.trim();
    let value = rest.strip_prefix("- ")?;
    Some(unquote_yaml_scalar(value.trim()))
}

/// 解析 `tags: a, b` 或 `tags: [a, b]`。
fn parse_inline_tag_list(raw: &str) -> Vec<String> {
    let trimmed = raw.trim();
    let inner = if trimmed.starts_with('[') && trimmed.ends_with(']') && trimmed.len() >= 2 {
        &trimmed[1..trimmed.len() - 1]
    } else {
        trimmed
    };

    inner
        .split(',')
        .map(|part| unquote_yaml_scalar(part.trim()))
        .filter(|part| !part.is_empty())
        .collect()
}

/// 去掉 YAML 标量两端的引号，保留标签原文。
fn unquote_yaml_scalar(value: &str) -> String {
    let trimmed = value.trim();
    if trimmed.len() >= 2
        && ((trimmed.starts_with('"') && trimmed.ends_with('"'))
            || (trimmed.starts_with('\'') && trimmed.ends_with('\'')))
    {
        return trimmed[1..trimmed.len() - 1].to_owned();
    }

    trimmed.to_owned()
}

/// 收集正文中所有独立 `#标签` 行，跳过 frontmatter 和代码围栏。
fn parse_hashtag_line_tags(content: &str) -> Vec<String> {
    let lines: Vec<&str> = content.split('\n').collect();
    let in_fence = code_fence_mask(&lines);
    let start = frontmatter_body_start_line(&lines);
    let mut tags = Vec::new();

    for (index, line) in lines.iter().enumerate().skip(start) {
        if in_fence[index] {
            continue;
        }

        if let Some(line_tags) = parse_hashtag_only_line(line) {
            tags.extend(line_tags);
        }
    }

    tags
}

/// 正文起始行：有闭合 frontmatter 时从结束标记之后开始。
fn frontmatter_body_start_line(lines: &[&str]) -> usize {
    if lines
        .first()
        .is_none_or(|line| line.trim_end_matches('\r').trim() != "---")
    {
        return 0;
    }

    for (index, line) in lines.iter().enumerate().skip(1) {
        if line.trim_end_matches('\r').trim() == "---" {
            return index + 1;
        }
    }

    0
}

/// 标记每一行是否位于 fenced code 内，围栏行本身也视为代码。
fn code_fence_mask(lines: &[&str]) -> Vec<bool> {
    let mut in_fence = false;
    let mut mask = Vec::with_capacity(lines.len());

    for line in lines {
        let trimmed = line.trim_end_matches('\r').trim();
        if trimmed.starts_with("```") || trimmed.starts_with("~~~") {
            mask.push(true);
            in_fence = !in_fence;
            continue;
        }

        mask.push(in_fence);
    }

    mask
}

/// 整行都是 `#标签` token 时返回这些标签，否则视为普通正文。
fn parse_hashtag_only_line(line: &str) -> Option<Vec<String>> {
    let trimmed = line.trim_end_matches('\r').trim();
    if trimmed.is_empty() || is_atx_heading(trimmed) {
        return None;
    }

    let mut tags = Vec::new();
    for token in trimmed.split_whitespace() {
        tags.push(hashtag_token_name(token)?);
    }

    if tags.is_empty() {
        None
    } else {
        Some(tags)
    }
}

/// CommonMark ATX 标题需要 `#` 后接空白；`#标签` 不是标题。
fn is_atx_heading(trimmed: &str) -> bool {
    let hashes = trimmed.chars().take_while(|ch| *ch == '#').count();
    if !(1..=6).contains(&hashes) {
        return false;
    }

    let rest = &trimmed[hashes..];
    rest.is_empty() || rest.starts_with(char::is_whitespace)
}

/// 解析单个 `#标签` token，拒绝 `##标题` 这种连续井号。
fn hashtag_token_name(token: &str) -> Option<String> {
    let rest = token.strip_prefix('#')?;
    if rest.starts_with('#') {
        return None;
    }

    normalize_tag_name(rest)
}

#[cfg(test)]
mod tests {
    use super::extract_tags;

    #[test]
    fn extracts_trailing_hashtags() {
        let content = "# 标题\n\n正文内容。\n\n#产品 #MVP #Agent\n";
        assert_eq!(extract_tags(content), vec!["Agent", "MVP", "产品"]);
    }

    #[test]
    fn extracts_frontmatter_list_and_inline_tags() {
        let list = "---\ntags:\n  - 研究\n  - 检索\n---\n\n# 标题\n";
        assert_eq!(extract_tags(list), vec!["检索", "研究"]);

        let inline = "---\ntags: [隐私, 架构]\n---\n\n正文\n";
        assert_eq!(extract_tags(inline), vec!["架构", "隐私"]);

        let csv = "---\ntags: 会议, 原型\n---\n\n正文\n";
        assert_eq!(extract_tags(csv), vec!["会议", "原型"]);
    }

    #[test]
    fn ignores_headings_paragraph_hashtags_and_code_fences() {
        let content =
            "# 标题\n\n段落里的 #看起来像标签 不是标签。\n\n```bash\n#todo\n```\n\n## 二级标题\n";
        assert!(extract_tags(content).is_empty());
    }

    #[test]
    fn removing_trailing_hashtags_clears_tags() {
        let with_tags = "# 标题\n\n正文\n\n#产品 #MVP\n";
        let without_tags = "# 标题\n\n正文\n";
        assert_eq!(extract_tags(with_tags), vec!["MVP", "产品"]);
        assert!(extract_tags(without_tags).is_empty());
    }

    #[test]
    fn unions_frontmatter_and_hashtag_lines() {
        let content = "---\ntags: [产品]\n---\n\n# 标题\n\n#会议\n";
        assert_eq!(extract_tags(content), vec!["产品", "会议"]);
    }

    #[test]
    fn hashtag_line_after_title_counts() {
        let content = "# 标题\n\n#产品 #MVP\n\n正文从这里开始。\n";
        assert_eq!(extract_tags(content), vec!["MVP", "产品"]);
    }
}
